File: src/tizenclaw/channel/mcp_client.cc
/*
 * Copyright (c) 2026 Samsung Electronics Co., Ltd All Rights Reserved
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "mcp_client.hh"
#include "../../common/logging.hh"

#include <fcntl.h>
#include <poll.h>
#include <sys/wait.h>
#include <unistd.h>
#include <iostream>
#include <thread>

namespace tizenclaw {

McpClient::McpClient(const std::string& server_name, const std::string& command,
                     const std::vector<std::string>& args, int timeout_ms)
    : server_name_(server_name), command_(command), args_(args), timeout_ms_(timeout_ms) {
    UpdateLastUsed();
}

McpClient::~McpClient() { Disconnect(); }

void McpClient::UpdateLastUsed() {
  last_used_ms_ = std::chrono::duration_cast<std::chrono::milliseconds>(
                      std::chrono::system_clock::now().time_since_epoch())
                      .count();
}

long long McpClient::GetLastUsedMs() const {
  return last_used_ms_;
}

bool McpClient::Connect() {
  if (is_connected_) return true;
  UpdateLastUsed();

  if (pipe(pipe_stdin_) == -1 || pipe(pipe_stdout_) == -1) {
    LOG(ERROR) << "MCP Client: Failed to create pipes for " << server_name_;
    return false;
  }

  // Make read end non-blocking
  int flags = fcntl(pipe_stdout_[0], F_GETFL, 0);
  fcntl(pipe_stdout_[0], F_SETFL, flags | O_NONBLOCK);

  pid_ = fork();
  if (pid_ < 0) {
    LOG(ERROR) << "MCP Client: Fork failed for " << server_name_;
    return false;
  }

  if (pid_ == 0) {
    // Child process: set up stdin/stdout
    dup2(pipe_stdin_[0], STDIN_FILENO);
    dup2(pipe_stdout_[1], STDOUT_FILENO);

    close(pipe_stdin_[1]);
    close(pipe_stdout_[0]);
    close(pipe_stdin_[0]);
    close(pipe_stdout_[1]);

    // Build args array
    std::vector<const char*> exec_args;
    exec_args.push_back(command_.c_str());
    for (const auto& arg : args_) {
      exec_args.push_back(arg.c_str());
    }
    exec_args.push_back(nullptr);

    execvp(command_.c_str(), const_cast<char* const*>(exec_args.data()));
    // If exec fails, exit child silently to avoid polluting parent logs
    _exit(1);
  }

  // Parent process
  close(pipe_stdin_[0]);
  close(pipe_stdout_[1]);
  is_connected_ = true;

  LOG(INFO) << "MCP Client: Process for " << server_name_ << " started (PID: " << pid_ << ")";

  // Perform 'initialize' handshake
  try {
    nlohmann::json init_params = {
        {"protocolVersion", "2024-11-05"},
        {"capabilities", nlohmann::json::object()},
        {"clientInfo", {{"name", "tizenclaw-mcp-client"}, {"version", "1.0.0"}}}
    };
    
    nlohmann::json response = SendRequestSync("initialize", init_params, 10000);
    if (response.contains("error")) {
      LOG(ERROR) << "MCP Client: Handshake failed for " << server_name_;
      Disconnect();
      return false;
    }
    
    // We must send notifications/initialized next
    nlohmann::json notif = {
      {"jsonrpc", "2.0"},
      {"method", "notifications/initialized"}
    };
    SendRpcMessage(notif);
    
    LOG(INFO) << "MCP Client: Handshake succeeded for " << server_name_;

  } catch (const std::exception& e) {
    LOG(ERROR) << "MCP Client: Initialization exception for " << server_name_ << ": " << e.what();
    Disconnect();
    return false;
  }

  return true;
}

void McpClient::Disconnect() {
  if (!is_connected_) return;

  is_connected_ = false;
  
  if (pipe_stdin_[1] != -1) close(pipe_stdin_[1]);
  if (pipe_stdout_[0] != -1) close(pipe_stdout_[0]);

  if (pid_ > 0) {
    kill(pid_, SIGTERM);
    int wstatus;
    int ret = 0;
    for (int i = 0; i < 10; ++i) {
      ret = waitpid(pid_, &wstatus, WNOHANG);
      if (ret > 0) break;
      std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }
    if (ret <= 0) {
      LOG(WARNING) << "MCP Client: " << server_name_ << " stuck, sending SIGKILL";
      kill(pid_, SIGKILL);
      waitpid(pid_, &wstatus, 0);
    }
    pid_ = -1;
  }
}

std::vector<McpClient::ToolInfo> McpClient::GetTools() {
  std::vector<ToolInfo> tools;
  if (!is_connected_) return tools;

  try {
    nlohmann::json response = SendRequestSync("tools/list", nlohmann::json::object(), 5000);
    if (response.contains("result") && response["result"].contains("tools")) {
      for (const auto& t : response["result"]["tools"]) {
        ToolInfo info;
        info.name = t.value("name", "");
        info.description = t.value("description", "");
        if (t.contains("inputSchema")) {
          info.input_schema = t["inputSchema"];
        }
        tools.push_back(info);
      }
    }
  } catch (const std::exception& e) {
    LOG(ERROR) << "MCP Client: GetTools error for " << server_name_ << ": " << e.what();
  }
  return tools;
}

nlohmann::json McpClient::CallTool(const std::string& tool_name, const nlohmann::json& arguments) {
  if (!is_connected_) return {{"error", "Not connected"}};

  nlohmann::json params = {
    {"name", tool_name},
    {"arguments", arguments}
  };

  UpdateLastUsed();

  try {
    nlohmann::json response = SendRequestSync("tools/call", params, timeout_ms_);
    LOG(INFO) << "MCP Client: CallTool response for " << tool_name << ": " << response.dump();
    if (response.contains("result")) {
      return response["result"];
    } else if (response.contains("error")) {
      LOG(WARNING) << "MCP Client: CallTool error from " << server_name_ << ": " << response["error"].dump();
      return {{"isError", true}, {"error", response["error"]}};
    }
    return {{"isError", true}, {"error", "Invalid response from server"}};
  } catch (const std::exception& e) {
    LOG(ERROR) << "MCP Client: CallTool exception for " << server_name_ << ": " << e.what();
    return {{"isError", true}, {"error", e.what()}};
  }
}

bool McpClient::SendRpcMessage(const nlohmann::json& message) {
  std::lock_guard<std::mutex> lock(io_mutex_);
  if (!is_connected_) return false;

  std::string data = message.dump() + "\n";
  ssize_t written = write(pipe_stdin_[1], data.c_str(), data.length());
  return written == static_cast<ssize_t>(data.length());
}

nlohmann::json McpClient::ReadRpcMessage(int timeout_ms) {
  if (!is_connected_) return nullptr;

  auto parse_line = [this]() -> nlohmann::json {
    size_t pos = read_buffer_.find('\n');
    if (pos != std::string::npos) {
      std::string line = read_buffer_.substr(0, pos);
      read_buffer_.erase(0, pos + 1);
      if (line.empty() || line == "\r") return nlohmann::json(nullptr);
      try {
        return nlohmann::json::parse(line);
      } catch (...) {
        return nlohmann::json(nullptr);
      }
    }
    return nlohmann::json(nullptr);
  };

  // Try parsing any full lines already in buffer
  nlohmann::json parsed = parse_line();
  if (!parsed.is_null()) return parsed;

  struct pollfd pfd;
  pfd.fd = pipe_stdout_[0];
  pfd.events = POLLIN;

  char chunk[4096];

  auto start_time = std::chrono::steady_clock::now();
  
  while (is_connected_) {
    auto now = std::chrono::steady_clock::now();
    int elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(now - start_time).count();
    int remaining_ms = std::max(1, timeout_ms - elapsed_ms);
    if (elapsed_ms >= timeout_ms) break; // Timeout

    int ret = poll(&pfd, 1, remaining_ms);
    if (ret < 0) {
      if (errno == EINTR) continue;
      break; 
    }
    if (ret == 0) break; // Timeout

    if (pfd.revents & POLLIN || pfd.revents & POLLHUP) {
      ssize_t bytes = read(pipe_stdout_[0], chunk, sizeof(chunk) - 1);
      if (bytes > 0) {
        chunk[bytes] = '\0';
        read_buffer_.append(chunk, bytes);
        
        parsed = parse_line();
        if (!parsed.is_null()) return parsed;
      } else if (bytes == 0) {
        // EOF
        LOG(WARNING) << "MCP Client: Server closed pipe for " << server_name_;
        Disconnect();
        break;
      } else {
        if (errno != EAGAIN && errno != EWOULDBLOCK) {
          Disconnect();
          break;
        }
      }
    }
  }
  return nlohmann::json(nullptr); // Timeout
}

nlohmann::json McpClient::SendRequestSync(const std::string& method, const nlohmann::json& params, int timeout_ms) {
  int req_id = next_req_id_++;
  nlohmann::json request = {
    {"jsonrpc", "2.0"},
    {"id", req_id},
    {"method", method},
    {"params", params}
  };

  if (!SendRpcMessage(request)) {
    throw std::runtime_error("Failed to send request");
  }

  // Poll loop: wait for ID match
  auto start_time = std::chrono::steady_clock::now();
  while (is_connected_) {
    auto now = std::chrono::steady_clock::now();
    int elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(now - start_time).count();
    int remaining_ms = std::max(1, timeout_ms - elapsed_ms);
    
    if (elapsed_ms >= timeout_ms) break;

    nlohmann::json res = ReadRpcMessage(remaining_ms);
    if (res.is_null()) continue; 
    if (res.is_discarded()) continue; // Parse error

    if (res.contains("id") && res["id"] == req_id) {
      return res;
    }
    
    // Non-matching IDs handled here
    if (res.contains("method")) {
      LOG(INFO) << "MCP Client: Notification from " << server_name_ << " -> " << res["method"];
    }
  }

  throw std::runtime_error("Server request timed out after " + std::to_string(timeout_ms) + "ms");
}

} // namespace tizenclaw

File: src/tizenclaw/src/channel/mcp_client.rs
//! MCP client — Model Context Protocol client for external tool servers.

use serde_json::{json, Value};
use std::io::{BufRead, BufReader, Write};
use std::process::{Child, Command, Stdio};
use std::sync::atomic::{AtomicI32, Ordering};
use std::sync::Mutex;
use std::time::{Duration, Instant};

use crate::llm::backend::LlmToolDecl;

pub type ToolDecl = LlmToolDecl;

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum McpTransport {
    Stdio { command: String, args: Vec<String> },
    Http { url: String },
}

pub struct McpClient {
    pub server_name: String,
    pub server_url: String,
    pub transport: McpTransport,
    timeout_ms: u64,
    child: Option<Child>,
    reader: Option<Mutex<BufReader<std::process::ChildStdout>>>,
    writer: Option<Mutex<std::process::ChildStdin>>,
    connected: bool,
    tools: Vec<ToolDecl>,
    next_req_id: AtomicI32,
    last_used_ms: u64,
}

impl McpClient {
    pub fn new(server_name: &str, command: &str, args: &[String], timeout_ms: u64) -> Self {
        Self {
            server_name: server_name.into(),
            server_url: format!("stdio://{}", server_name),
            transport: McpTransport::Stdio {
                command: command.into(),
                args: args.to_vec(),
            },
            timeout_ms,
            child: None,
            reader: None,
            writer: None,
            connected: false,
            tools: Vec::new(),
            next_req_id: AtomicI32::new(1),
            last_used_ms: Self::now_ms(),
        }
    }

    pub fn connect(config: &Value) -> Result<Self, String> {
        let transport_name = config
            .get("transport")
            .and_then(Value::as_str)
            .unwrap_or("stdio")
            .trim()
            .to_ascii_lowercase();
        let server_name = config
            .get("name")
            .and_then(Value::as_str)
            .unwrap_or("mcp")
            .trim();
        let timeout_ms = config
            .get("timeout_ms")
            .and_then(Value::as_u64)
            .unwrap_or(30_000);

        let mut client = match transport_name.as_str() {
            "http" => {
                let url = config
                    .get("url")
                    .or_else(|| config.get("server_url"))
                    .and_then(Value::as_str)
                    .map(str::trim)
                    .filter(|value| !value.is_empty())
                    .ok_or_else(|| "mcp http transport requires url".to_string())?;
                Self {
                    server_name: server_name.to_string(),
                    server_url: url.to_string(),
                    transport: McpTransport::Http {
                        url: url.to_string(),
                    },
                    timeout_ms,
                    child: None,
                    reader: None,
                    writer: None,
                    connected: false,
                    tools: Vec::new(),
                    next_req_id: AtomicI32::new(1),
                    last_used_ms: Self::now_ms(),
                }
            }
            _ => {
                let command = config
                    .get("command")
                    .and_then(Value::as_str)
                    .map(str::trim)
                    .filter(|value| !value.is_empty())
                    .ok_or_else(|| "mcp stdio transport requires command".to_string())?;
                let args = config
                    .get("args")
                    .and_then(Value::as_array)
                    .map(|items| {
                        items
                            .iter()
                            .filter_map(|value| value.as_str().map(str::to_string))
                            .collect::<Vec<_>>()
                    })
                    .unwrap_or_default();
                Self::new(server_name, command, &args, timeout_ms)
            }
        };

        if client.open() {
            Ok(client)
        } else {
            Err(format!("Failed to connect MCP client '{}'", server_name))
        }
    }

    fn now_ms() -> u64 {
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis() as u64
    }

    fn update_last_used(&mut self) {
        self.last_used_ms = Self::now_ms();
    }

    pub fn last_used_ms(&self) -> u64 {
        self.last_used_ms
    }

    pub fn open(&mut self) -> bool {
        if self.connected {
            return true;
        }
        self.update_last_used();

        match &self.transport {
            McpTransport::Stdio { command, args } => {
                let mut child = match Command::new(command)
                    .args(args)
                    .stdin(Stdio::piped())
                    .stdout(Stdio::piped())
                    .stderr(Stdio::null())
                    .spawn()
                {
                    Ok(child) => child,
                    Err(err) => {
                        log::error!(
                            "MCP client '{}' failed to spawn transport: {}",
                            self.server_name,
                            err
                        );
                        return false;
                    }
                };

                let pid = child.id();
                let Some(stdout) = child.stdout.take() else {
                    return false;
                };
                let Some(stdin) = child.stdin.take() else {
                    return false;
                };

                self.reader = Some(Mutex::new(BufReader::new(stdout)));
                self.writer = Some(Mutex::new(stdin));
                self.child = Some(child);
                self.connected = true;

                log::debug!("MCP client '{}' started (pid {})", self.server_name, pid);
            }
            McpTransport::Http { .. } => {
                self.connected = true;
            }
        }

        let init_params = json!({
            "protocolVersion": "2024-11-05",
            "capabilities": {},
            "clientInfo": {"name": "tizenclaw-mcp-client", "version": "1.0.0"}
        });

        match self.rpc_request("initialize", &init_params, 10_000) {
            Ok(resp) if resp.get("error").is_none() => {}
            Ok(_) | Err(_) => {
                self.disconnect();
                return false;
            }
        }

        let _ = self.send_notification("notifications/initialized", &json!({}));
        true
    }

    pub fn disconnect(&mut self) {
        self.connected = false;
        self.reader = None;
        self.writer = None;

        if let Some(mut child) = self.child.take() {
            let _ = child.kill();
            let _ = child.wait();
        }
    }

    pub fn discover_tools(&mut self) -> &[ToolDecl] {
        if !self.connected {
            return &self.tools;
        }

        match self.rpc_request("tools/list", &json!({}), 5_000) {
            Ok(resp) => {
                let tools = resp
                    .get("result")
                    .and_then(|value| value.get("tools"))
                    .and_then(Value::as_array)
                    .cloned()
                    .unwrap_or_default();
                self.tools = tools
                    .iter()
                    .filter_map(|tool| {
                        let name = tool.get("name").and_then(Value::as_str)?;
                        Some(ToolDecl {
                            name: format!("mcp_{}_{}", self.server_name, name),
                            description: tool
                                .get("description")
                                .and_then(Value::as_str)
                                .unwrap_or("")
                                .to_string(),
                            parameters: tool
                                .get("inputSchema")
                                .cloned()
                                .unwrap_or_else(|| json!({"type":"object"})),
                        })
                    })
                    .collect();
            }
            Err(err) => {
                log::warn!("MCP client '{}' tools/list failed: {}", self.server_name, err);
            }
        }

        &self.tools
    }

    pub fn list_tools(&mut self) -> Vec<ToolDecl> {
        self.discover_tools();
        self.tools.clone()
    }

    pub fn get_tools(&self) -> &[ToolDecl] {
        &self.tools
    }

    pub fn is_connected(&self) -> bool {
        self.connected
    }

    pub fn call_tool(&mut self, tool_name: &str, arguments: &Value) -> Result<Value, String> {
        if !self.connected {
            return Err("Not connected".into());
        }
        self.update_last_used();

        let response = self.rpc_request(
            "tools/call",
            &json!({"name": tool_name, "arguments": arguments}),
            self.timeout_ms,
        )?;
        if let Some(result) = response.get("result") {
            Ok(result.clone())
        } else if let Some(error) = response.get("error") {
            Err(error.to_string())
        } else {
            Err("Invalid MCP response".to_string())
        }
    }

    fn send_notification(&self, method: &str, params: &Value) -> Result<(), String> {
        let message = json!({
            "jsonrpc": "2.0",
            "method": method,
            "params": params
        });

        match &self.transport {
            McpTransport::Stdio { .. } => self.send_stdio_message(&message),
            McpTransport::Http { url } => {
                tokio::runtime::Builder::new_current_thread()
                    .enable_all()
                    .build()
                    .map_err(|err| err.to_string())?
                    .block_on(async {
                        reqwest::Client::builder()
                            .timeout(Duration::from_secs(10))
                            .build()
                            .map_err(|err| err.to_string())?
                            .post(url)
                            .header("Content-Type", "application/json")
                            .body(message.to_string())
                            .send()
                            .await
                            .map(|_| ())
                            .map_err(|err| err.to_string())
                    })
            }
        }
    }

    fn rpc_request(&self, method: &str, params: &Value, timeout_ms: u64) -> Result<Value, String> {
        let req_id = self.next_req_id.fetch_add(1, Ordering::SeqCst);
        let request = json!({
            "jsonrpc": "2.0",
            "id": req_id,
            "method": method,
            "params": params
        });

        match &self.transport {
            McpTransport::Stdio { .. } => {
                self.send_stdio_message(&request)?;
                self.read_stdio_response(req_id, timeout_ms)
            }
            McpTransport::Http { url } => {
                tokio::runtime::Builder::new_current_thread()
                    .enable_all()
                    .build()
                    .map_err(|err| err.to_string())?
                    .block_on(async {
                        let response = reqwest::Client::builder()
                            .timeout(Duration::from_millis(timeout_ms))
                            .build()
                            .map_err(|err| err.to_string())?
                            .post(url)
                            .header("Content-Type", "application/json")
                            .body(request.to_string())
                            .send()
                            .await
                            .map_err(|err| err.to_string())?;
                        response.json::<Value>().await.map_err(|err| err.to_string())
                    })
            }
        }
    }

    fn send_stdio_message(&self, message: &Value) -> Result<(), String> {
        let writer = self.writer.as_ref().ok_or("No writer")?;
        let mut writer = writer.lock().map_err(|err| err.to_string())?;
        writer
            .write_all(format!("{}\n", message).as_bytes())
            .map_err(|err| err.to_string())?;
        writer.flush().map_err(|err| err.to_string())
    }

    fn read_stdio_response(&self, request_id: i32, timeout_ms: u64) -> Result<Value, String> {
        let reader = self.reader.as_ref().ok_or("No reader")?;
        let mut reader = reader.lock().map_err(|err| err.to_string())?;
        let start = Instant::now();
        let timeout = Duration::from_millis(timeout_ms);
        let mut line = String::new();

        loop {
            if start.elapsed() >= timeout {
                return Err(format!("Timeout after {}ms", timeout_ms));
            }

            line.clear();
            match reader.read_line(&mut line) {
                Ok(0) => return Err("EOF".into()),
                Ok(_) => {
                    let trimmed = line.trim();
                    if trimmed.is_empty() {
                        continue;
                    }
                    let response: Value =
                        serde_json::from_str(trimmed).map_err(|err| err.to_string())?;
                    if response.get("id").and_then(Value::as_i64) == Some(request_id as i64) {
                        return Ok(response);
                    }
                }
                Err(err) => return Err(err.to_string()),
            }
        }
    }
}

impl Drop for McpClient {
    fn drop(&mut self) {
        self.disconnect();
    }
}

pub struct McpClientManager {
    clients: Vec<McpClient>,
}

impl Default for McpClientManager {
    fn default() -> Self {
        Self::new()
    }
}

impl McpClientManager {
    pub fn new() -> Self {
        Self { clients: Vec::new() }
    }

    pub fn load_config_and_connect(&mut self, path: &str) -> bool {
        let Ok(content) = std::fs::read_to_string(path) else {
            return false;
        };
        let Ok(config) = serde_json::from_str::<Value>(&content) else {
            return false;
        };

        if let Some(servers) = config.get("servers").and_then(Value::as_array) {
            for server in servers {
                let Ok(mut client) = McpClient::connect(server) else {
                    continue;
                };
                client.discover_tools();
                self.clients.push(client);
            }
        }

        !self.clients.is_empty()
    }

    pub fn get_all_tools(&self) -> Vec<ToolDecl> {
        self.clients
            .iter()
            .flat_map(|client| client.get_tools().to_vec())
            .collect()
    }

    pub fn call_tool(&mut self, full_name: &str, args: &Value) -> Option<Value> {
        for client in &mut self.clients {
            let prefix = format!("mcp_{}_", client.server_name);
            if let Some(tool_name) = full_name.strip_prefix(&prefix) {
                return Some(match client.call_tool(tool_name, args) {
                    Ok(value) => value,
                    Err(err) => json!({"isError": true, "error": err}),
                });
            }
        }
        None
    }
}

#[cfg(test)]
mod tests {
    use super::McpClient;
    use serde_json::json;
    use std::fs;
    use tempfile::tempdir;

    #[test]
    fn stdio_connect_and_list_tools() {
        let dir = tempdir().expect("tempdir");
        let script_path = dir.path().join("fake_mcp.py");
        fs::write(
            &script_path,
            r#"#!/usr/bin/env python3
import json
import sys
for line in sys.stdin:
    line = line.strip()
    if not line:
        continue
    req = json.loads(line)
    method = req.get("method")
    if method == "initialize":
        print(json.dumps({
            "jsonrpc": "2.0",
            "id": req.get("id"),
            "result": {"protocolVersion": "2024-11-05", "capabilities": {"tools": {}}}
        }), flush=True)
    elif method == "notifications/initialized":
        continue
    elif method == "tools/list":
        print(json.dumps({
            "jsonrpc": "2.0",
            "id": req.get("id"),
            "result": {
                "tools": [{
                    "name": "echo",
                    "description": "Echo a payload",
                    "inputSchema": {"type": "object"}
                }]
            }
        }), flush=True)
    elif method == "tools/call":
        print(json.dumps({
            "jsonrpc": "2.0",
            "id": req.get("id"),
            "result": {"content": [{"type": "text", "text": "ok"}]}
        }), flush=True)
"#,
        )
        .expect("write fake server");

        let mut client = McpClient::connect(&json!({
            "name": "fake",
            "transport": "stdio",
            "command": "python3",
            "args": [script_path.to_string_lossy().to_string()]
        }))
        .expect("connect mcp");

        let tools = client.list_tools();
        assert_eq!(tools.len(), 1);
        assert_eq!(tools[0].name, "mcp_fake_echo");

        let result = client.call_tool("echo", &json!({"text": "hello"})).expect("call tool");
        assert_eq!(result["content"][0]["text"], "ok");
    }
}
